### PR TITLE
Add search filter when displaying namespaces Fixes #15 and #49

### LINKED ISSUE #15: Quick jump a namespaces ?
can I jump a namespaces quickly?

### LINKED ISSUE #49: Search filter in namespace modal dialog
Hello!

Please add a filter to the namespace modal selection window.

Currently awfully inconvenient to choose the necessary namespace when there are more than 100...

I would be very grateful!

---

I was thinking of having a quick search filter in the namespaces modal. Would that work for you?

_Originally posted by @astefanutti in https://github.com/astefanutti/kubebox/issues/15#issuecomment-402968374_

### PR DESCRIPTION
**Pull request by johnpoth**
Fixes #15 and fixes #49.

Adds search bar to namespace selection. Search is triggered when user starts typing. Further improvements might include highlighting match.

Thanks!

### CODE REVIEW

File: lib/ui/namespaces.js [hunk 1]
```diff
@@ -32,6 +32,114 @@ function namespaces_list(screen) {
...
+    const len = namespaces_list.items.length;
+    for(i = mod(i,len); i < len && i >= 0 ; i = mod(i+1,len)){
+        if(!namespaces_list.items[i].hidden) { 
+            namespaces_list.selected = i;
+            namespaces_list.scrollTo(namespaces_list.items[i].position.top);
+            screen.render();
+            return;
+        }
+    }
+  };
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: It seems these two methods could be factorised.

File: lib/ui/namespaces.js [hunk 2]
```diff
@@ -32,6 +32,114 @@ function namespaces_list(screen) {
...
+    inputOnFocus : true,
+    style : {
+      fg       : 'white',
+      label    : { bold: true },
+      border   : { fg: 'white' },
+      selected : { bg: 'blue' },
+    }
+  });
+
+  search.__oolistener = search._listener;
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: I would propose to add a comment that this is the canonical way of extending component in Blessed.

**Comment by astefanutti** [comment]
I've thought a bit about the best way to integrate the filter and come up with:
Kubebox  - namespaces filter
It's a bit more UI work, but that would solve some of the issues mentioned above. Besides, it'd make obvious for users that filtering is possible.

WDYT?

**Comment by johnpoth** [comment]
Excellent feedback ! Let me work some magic and update this PR

**Comment by astefanutti** [comment]
Cool, I think we'll be able to generalise the work done for the namespaces and reuse it so that we can provide filtering for other widgets like pods, logs, ...

**Comment by johnpoth** [comment]
PR updated, lint errors may apply

**Comment by astefanutti** [comment]
It seems the PR hasn't picked up the changes. I think the PR was created from your fork, while you've pushed the branch to this repository. Would you mind pushing to your fork as well?

**Comment by johnpoth** [comment]
PR updated

File: lib/ui/namespaces.js [hunk 3]
```diff
@@ -6,13 +6,24 @@ const blessed = require('blessed'),
...
 function namespaces_list(screen) {
   // TODO: display a list table with some high level info about the namespaces
+  const namespaces_box = blessed.box({
+    parent : screen,
+    top    : 'center',
+    left   : 'center',
+    width  : '50%',
+    height : '50%',
+    label  : 'Namespaces',
+    border : { type: 'line' }
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: nit: missing trailing comma.

File: lib/ui/namespaces.js [hunk 4]
```diff
@@ -6,13 +6,24 @@ const blessed = require('blessed'),
...
+    parent    : namespaces_box,
+    height    : 'shrink',
     left      : 'center',
-    width     : '50%',
-    height    : '50%',
-    label     : 'Namespaces',
+    bottom    :  0,
+    align     : 'left',
+    top       : 3,
+    width     : '97%',
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: To be replaced with `100%-2`.

File: lib/ui/namespaces.js [hunk 5]
```diff
@@ -6,13 +6,24 @@ const blessed = require('blessed'),
...
+    label  : 'Namespaces',
+    border : { type: 'line' }
+  });
+
   const namespaces_list = blessed.list({
-    screen    : screen,
-    top       : 'center',
+    parent    : namespaces_box,
+    height    : 'shrink',
     left      : 'center',
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: Can be removed.

File: lib/ui/namespaces.js [hunk 6]
```diff
@@ -32,12 +43,104 @@ function namespaces_list(screen) {
     }
   });
 
-  return namespaces_list;
+  blessed.text({
+    parent  : namespaces_box,
+    width   : 9,
+    left    : 3,
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: Should be `1`.

File: lib/ui/namespaces.js [hunk 7]
```diff
@@ -32,12 +43,104 @@ function namespaces_list(screen) {
...
+    top     : 1,
+    align   : 'left',
+    content : 'Filter :'
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: nit: missing trailing comma.

File: lib/ui/namespaces.js [hunk 8]
```diff
@@ -32,12 +43,104 @@ function namespaces_list(screen) {
...
+  }); 
+
+  const search = blessed.textbox({
+    parent       : namespaces_box,
+    border       : 'line',
+    width        : '85%',
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: Should be replaced with `100%-11`.

File: lib/ui/namespaces.js [hunk 9]
```diff
@@ -32,12 +43,104 @@ function namespaces_list(screen) {
...
+    height       : 3,
+    top          : 0,
+    right        : 1,
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: Should be `0`.

File: lib/ui/namespaces.js [hunk 10]
```diff
@@ -32,12 +43,104 @@ function namespaces_list(screen) {
...
+      label    : { bold: true },
+      border   : { fg: 'white' },
+      selected : { bg: 'blue' },
+    }
+  });
+
+  // Canonical way of extending components in Blessed
+  search.__oolistener = search._listener;
+  search._listener = function(ch, key) {
+    const word = ['up', 'down','enter'];
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: nit: missing extra space.

File: lib/ui/namespaces.js [hunk 11]
```diff
@@ -32,12 +43,104 @@ function namespaces_list(screen) {
...
+            screen.render();
+            return;
+        }
+    }
+  }
+
+  function mod(a, b) {
+    return (((a)%b)+b)%b;
+  }
+  return { 'search': search, 'box': namespaces_box, 'list': namespaces_list };
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: Can be replaced with `{ search, box, list } `.

File: lib/ui/namespaces.js [hunk 12]
```diff
@@ -32,12 +43,104 @@ function namespaces_list(screen) {
...
+  search._listener = function(ch, key) {
+    const word = ['up', 'down','enter'];    
+    if (word.includes(key.name)) {
+        return namespaces_list.emit('keypress', ch, key);
+    }
+    const ret = this.__oolistener(ch, key);
+    let i = 0;
+    namespaces_list.items.forEach((element, pos) => {
+        if (!element.getContent().includes(search.value)){
+            element.position.top = 0;
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: Is that really needed?

**Comment by johnpoth** [comment]
@astefanutti I've push some changes to address some sizing issues I had, I'll take a look at your comments ASAP

File: lib/ui/namespaces.js (hunk 12 above)
Comment (johnpoth) [reply lib/ui/namespaces.js]: Removed

File: lib/ui/namespaces.js (hunk 11 above)
Comment (johnpoth) [reply lib/ui/namespaces.js]: Replaced

File: lib/ui/namespaces.js (hunk 10 above)
Comment (johnpoth) [reply lib/ui/namespaces.js]: done.

File: lib/ui/namespaces.js (hunk 9 above)
Comment (johnpoth) [reply lib/ui/namespaces.js]: done.

File: lib/ui/namespaces.js (hunk 8 above)
Comment (johnpoth) [reply lib/ui/namespaces.js]: done.

File: lib/ui/namespaces.js (hunk 7 above)
Comment (johnpoth) [reply lib/ui/namespaces.js]: done.

File: lib/ui/namespaces.js (hunk 6 above)
Comment (johnpoth) [reply lib/ui/namespaces.js]: done.

File: lib/ui/namespaces.js (hunk 5 above)
Comment (johnpoth) [reply lib/ui/namespaces.js]: done.

File: lib/ui/namespaces.js [hunk 13]
```diff
@@ -32,12 +41,103 @@ function namespaces_list(screen) {
...
+    width        : '100%-11',
+    height       : 3,
+    top          : 0,
+    right        : 0,
+    keys         : true,
+    mouse        : true,
+    inputOnFocus : true,
+    style : {
+      fg       : 'white',
+      label    : { bold: true },
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: I think that belongs to the `box` now.

File: lib/ui/namespaces.js [hunk 14]
```diff
@@ -32,12 +41,103 @@ function namespaces_list(screen) {
...
+      border   : { fg: 'white' },
+      selected : { bg: 'blue' },
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: I think that belongs the the `list` now.

File: lib/ui/namespaces.js [hunk 15]
```diff
@@ -32,12 +41,103 @@ function namespaces_list(screen) {
...
+
+  const search = blessed.textbox({
+    parent       : box,
+    border       : 'line',
+    width        : '100%-11',
+    height       : 3,
+    top          : 0,
+    right        : 0,
+    keys         : true,
+    mouse        : true,
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: I think that belongs to the `list` now.

File: lib/ui/namespaces.js [hunk 16]
```diff
@@ -6,13 +6,22 @@ const blessed = require('blessed'),
 
 function namespaces_list(screen) {
```
Comment (astefanutti) [inline lib/ui/namespaces.js]: I'd suggest to rename it to `namespaces_ui`.

**Comment by johnpoth** [comment]
Should be ok! Thanks !

**Comment by astefanutti** [comment]
Awesome! Thanks!

**Pull request merged**